File: backend/app/schemas/catalog.py
from pydantic import BaseModel, Field


class CatalogPricingRead(BaseModel):
    mode: str
    amount_minor: int | None = None
    currency: str | None = None
    display_price: bool
    can_add_to_cart: bool
    can_checkout_online: bool
    action: str
    action_label: str


class CatalogImageRead(BaseModel):
    path: str
    alt_text: str


class CatalogSpecificationRead(BaseModel):
    spec_key: str
    label: str
    value_text: str
    unit: str | None = None


class CatalogVariantRead(BaseModel):
    id: str
    display_name: str
    sku: str
    option_values: dict[str, str]


class CatalogDocumentRead(BaseModel):
    title: str
    document_type: str
    path: str
    content_type: str
    version: str


class CatalogProductRead(BaseModel):
    id: str
    name: str
    slug: str
    sku: str
    description: str
    product_family: str | None
    manufacturer: str
    category: str
    public_path: str
    primary_image: CatalogImageRead | None
    pricing: CatalogPricingRead


class CatalogProductDetailRead(CatalogProductRead):
    images: list[CatalogImageRead] = Field(default_factory=list)
    variants: list[CatalogVariantRead] = Field(default_factory=list)
    documents: list[CatalogDocumentRead] = Field(default_factory=list)
    specifications: list[CatalogSpecificationRead] = Field(
        default_factory=list,
    )


class CatalogProductListResponse(BaseModel):
    products: list[CatalogProductRead] = Field(default_factory=list)

File: frontend/src/api/catalog.ts
export type CatalogPricing = {
  mode: string;
  amount_minor: number | null;
  currency: string | null;
  display_price: boolean;
  can_add_to_cart: boolean;
  can_checkout_online: boolean;
  action: string;
  action_label: string;
};

export type CatalogImage = {
  path: string;
  alt_text: string;
};

export type CatalogSpecification = {
  spec_key: string;
  label: string;
  value_text: string;
  unit: string | null;
};

export type CatalogVariant = {
  id: string;
  display_name: string;
  sku: string;
  option_values: Record<string, string>;
};

export type CatalogDocument = {
  title: string;
  document_type: string;
  path: string;
  content_type: string;
  version: string;
};

export type CatalogProduct = {
  id: string;
  name: string;
  slug: string;
  sku: string;
  description: string;
  product_family: string | null;
  manufacturer: string;
  category: string;
  public_path: string;
  primary_image: CatalogImage | null;
  pricing: CatalogPricing;
};

export type CatalogProductDetail =
  CatalogProduct & {
    images: CatalogImage[];
    variants: CatalogVariant[];
    documents: CatalogDocument[];
    specifications: CatalogSpecification[];
  };

type ProductListPayload = {
  products: CatalogProduct[];
};

async function readError(
  response: Response,
): Promise<string> {
  try {
    const payload = (await response.json()) as {
      detail?: string;
    };

    if (
      typeof payload.detail === "string"
    ) {
      return payload.detail;
    }
  } catch {
    // Fall through.
  }

  return (
    `Catalog request failed: ${response.status}`
  );
}

export async function getCatalogProducts(): Promise<
  CatalogProduct[]
> {
  const response = await fetch(
    "/api/catalog/products",
    {
      cache: "no-store",
      credentials: "include",
    },
  );

  if (!response.ok) {
    throw new Error(
      await readError(response),
    );
  }

  const payload =
    (await response.json()) as ProductListPayload;

  return payload.products;
}

export async function getCatalogProduct(
  slug: string,
): Promise<CatalogProductDetail> {
  const response = await fetch(
    `/api/catalog/products/${encodeURIComponent(slug)}`,
    {
      cache: "no-store",
      credentials: "include",
    },
  );

  if (!response.ok) {
    throw new Error(
      await readError(response),
    );
  }

  return response.json() as Promise<
    CatalogProductDetail
  >;
}
